pandas/core/arrays/datetimelike.py

```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
             return self._box_values(self.asi8)
         return super(DatetimeLikeArrayMixin, self).astype(dtype, copy)
 
+    # ------------------------------------------------------------------
+    # ExtensionArray Interface
+    # TODO:
+    #   * _from_sequence
+    #   * argsort / _values_for_argsort
+    #   * _reduce
+
+    def unique(self):
+        result = unique1d(self.asi8)
+        return type(self)(result, dtype=self.dtype)
+
+    def _validate_fill_value(self, fill_value):
+        """
+        If a fill_value is passed to `take` convert it to an i8 representation,
+        raising ValueError if this is not possible.
+
+        Parameters
+        ----------
+        fill_value : object
+
+        Returns
+        -------
+        fill_value : np.int64
+
+        Raises
+        ------
+        ValueError
+        """
+        raise AbstractMethodError(self)
+
+    def take(self, indices, allow_fill=False, fill_value=None):
+        if allow_fill:
+            fill_value = self._validate_fill_value(fill_value)
+
+        new_values = take(self.asi8,
+                          indices,
+                          allow_fill=allow_fill,
+                          fill_value=fill_value)
+
+        return type(self)(new_values, dtype=self.dtype)
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        # for TimedeltaArray and PeriodArray; DatetimeArray overrides
+        freqs = {x.freq for x in to_concat}
+        assert len(freqs) == 1
+        freq = list(freqs)[0]
+        values = np.concatenate([x.asi8 for x in to_concat])
+        return cls._simple_new(values, freq=freq)
+
+    def copy(self, deep=False):
+        values = self.asi8
+        if deep:
```

Comment:
@jorisvandenbossche does this look like the correct behavior to you? I was thinking that `.copy()` should break a chain such that in

```python
a = DatetimeArray(...)
b = a.copy()
a[0] = np.nan
```

`b[0]` should not be updated along with `a[0]`. IIUC, with the current implementation `a` and `b` would have the same underlying ndarray (with `deep=False`).

In my head, the `deep` parameter is really only useful when the "scalars" are actually containers (nested data). In this case, `deep=True` would mean recursively copy the values in the "scalar" containers.

Reply:
The `if deep:` branch here was based on jeff's comment [here](https://github.com/pandas-dev/pandas/pull/23415#discussion_r232497536).  I haven't thought about this enough to have a strong opinion about it other than that we shouldn't silently ignore the `deep` kwarg.

Reply:
Or simply remove `deep` from the interface? (but that can maybe discussed elsewhere, as it is already part of the interface currenlty)

In any case, I think copy, even with deep=False (the default), should clearly copy the values so that modifying the copy does not change the original one like Tom showed in the example.

Reply:
We have https://github.com/pandas-dev/pandas/issues/22314 for the meaning of `deep`.